data/code/lib.c

```diff
@@ -216,3 +217,185 @@ int itemshufflemgr_calcslot_ex(const unsigned int* const itemshufflemgr,
+        );
+
+    ++sectionCount;
+}
+
+void set_force_lap()
+{
+    register char reg0 asm ("r0");
+    register char reg9 asm ("r9");
+
+    bool isSection;
+
+    passed = (char)0;
+    
+    // setpass will have already run by this point
+    asm("rlwinm  %r0,%r3,0x0,0x18,0x1f     \t\n" // r27 = (char)r3
+        "or %r9, %r0, %r0                  \t\n" // store r0 in r6
+        );
+
+    passed = reg9;
+
+    isSection = false;
+
+    // is the "section point" bit set?
+    asm("lwz %r3, 0x0044 (%r29)            \t\n" // r3 = Sector data at *(r30 + 0x44)
+        "lwz %r3, 0x0008 (%r3)             \t\n" // r3 = *(r3 + 8) (first check data)
+        "lbz %r3, 0x001B (%r3)             \t\n" // r3 = *(r3 + 0x1B) ("section checkpoint" flag)
+        "subic %r0, %r3, 0x01	           \t\n"
+        "subfe %r3, %r0, %r3	           \t\n"
+        "rlwinm  %r0, %r3, 0x0, 0x18, 0x1f \t\n" // cast to byte
+        );
+
+    if (reg0 != '\0')
+        isSection = true;
+
+    if (isSection && (passed))
+    {
+        force_lap_increment();
+    }
+
+    // since we hijacked an instruction line, run this so MKDD doesn't cry
+    asm("lwz        %r3,0x3c(%r29) \t\n");
+}
+
+// TODO: debug version of all of this
+void force_lap_increment()
+{
+    register unsigned int reg0 asm ("r0");
+    register unsigned int reg3 asm ("r3");
+    register unsigned int *kartcheck asm ("r29");
+
+    asm("lwz        %r3,0x2c(%r29)           \t\n"); // get lap count
+
+    if ((int)reg3 < 0)
+    {
+        //OSReport("Lap count < 0! Dodging assert failure...");
+
+        asm("or         %r3,%r29,%r29            \t\n");
+        lap_increment();
+    }
+
+    asm("or         %r3,%r29,%r29            \t\n"
+        "lbz        %r3,0x29(%r3)            \t\n" // isGoal
+        "rlwinm     %r0,%r3,0x0,0x18,0x1f    \t\n"
+        );
+
+    if (reg0 == 0)
+        KartChecker__setLapTime(kartcheck);
+
+    asm("li		%r0,0x1                  \t\n"
+        "or     %r3,%r29,%r29            \t\n"
+        "stb	%r0,0x28(%r29)           \t\n"
+        "or     %r3,%r29,%r29            \t\n"
+        );
+
+    lap_increment();
+
+    asm("or         %r3,%r29,%r29            \t\n"
+        "lbz        %r3,0x29(%r3)            \t\n" // isGoal
+        "rlwinm     %r0,%r3,0x0,0x18,0x1f    \t\n"
+        );
+
+    if ((reg0 & 0xff) == 0)
+    {
+        asm("lwz        %r3,0x2c(%r29)           \t\n"
+            "lwz        %r0,0xc(%r29)            \t\n"
+            );
+
+        if (reg0 <= reg3)
+        {
+            // setGoal and setGoalTime
+            do_goal_routine();
+        }
+    }
+}
+
+// redundant lap/goal funcs because it doesn't exist in retail
+void lap_increment()
+{ 
+    asm("lwz        %r4,0x2c(%r3) \t\n"
+        "lwz        %r0,0xc(%r3)  \t\n"
+        "cmpw       %r4,%r0       \t\n"
+        "bgelr                    \t\n"
+        "addi       %r0,%r4,0x1   \t\n"
+        "stw        %r0,0x2c(%r3) \t\n"
+        "blr                      \t\n"
```

Comment:
I _think_ this final `blr` instruction would be a duplicate, as `lap_increment()` would have its own `blr` instruction.